numba/core/typing/arraydecl.py

```diff
@@ -503,7 +504,15 @@ def generic_resolve(self, ary, attr):
         # Resolution of other attributes, for record arrays
         if isinstance(ary.dtype, types.Record):
             if attr in ary.dtype.fields:
-                return ary.copy(dtype=ary.dtype.typeof(attr), layout='A')
+                attr_dtype = ary.dtype.typeof(attr)
+                if isinstance(attr_dtype, types.NestedArray):
+                    return ary.copy(
+                        dtype=attr_dtype.dtype,
+                        ndim=ary.ndim + attr_dtype.ndim,
```

Comment:
@stuartarchibald I think you previously expressed some concern to me OOB that the addition of the dimensions of the "outer" array and the nested array were being summed to create the new dimension of the array. However, this makes sense to me. For example:

```python
import numpy as np

dtype = np.dtype([
    ('x', np.float32),
    ('y', np.float32, (4, 2))
])

x = np.zeros((5, 7), dtype=dtype)
print(len(x['y'].shape))
```

prints `4`, because getting an item of an array of records where the item is a nested array "slices" the array such that you get an array of all the nested arrays (the shape of `x['y']` here is `(5, 7, 4, 2)` - the dimensions of the array concatenated with the dimensions of the nested array).

Reply:
Thanks for the explanation @gmarkall, as noted at the public meeting, my concern was that lowering was working "by accident", however, following the above, I think it does make sense, and also explains why slicing etc works ok.